### PR TITLE
GBM: use windowing method for HDR activation

### PR DESCRIPTION
**Pull request by lrusak**
This changes the way the HDR activation is done when using GBM to use the interface we have in `CWinSystemBase`.

This is needed to allow HDR activation when not using direct-to-plane rendering.

This also enables the use of the new HDR setting in `settings -> player -> video -> processing -> Use display HDR capabilities`.

The problem here is that we can't yet actually detect if the monitor is HDR capable or not. We need the detection that is in #19141. Currently we just detect if we can output HDR metadata (if the GPU/connector supports it).

I've also added an extra commit to add some debug logging for the HDR metadata to the `VIDEO` component logging.

For platforms that were able to use direct-to-plane and HDR nothing should change. For regular video rendering this will work through the normal renderer `CLinuxRendererGLES` (but not `CRendererDRMPRIMEGLES` yet or `CLinuxRendererGL`) but you may have to disable the new setting to use tone mapping if you don't actually have an HDR capable display.

cc @jernejsk and @popcornmix for testing on their devices :chart_with_upwards_trend:

### CODE REVIEW

**Review by popcornmix — APPROVED** [review]
Tested on Pi4. HDR colours correct when "Use display HDR" is enabled and incorrect when disabled.

File: xbmc/windowing/gbm/WinSystemGbm.h
Comment (garbear) [inline xbmc/windowing/gbm/WinSystemGbm.h:23]: To keep platform-specific code out of headers, can this be moved into the cpp file? Then maybe forward-declare here.

```c++
struct hdr_metadata_infoframe;
struct hdr_output_metadata;
```

Comment (garbear) [reply xbmc/windowing/gbm/WinSystemGbm.h:23]: The structs can also be used to a "helper header" like `xbmc/windowing/gbm/GbmTypes.h` but they're probably not used elsewhere.

File: xbmc/windowing/gbm/WinSystemGbm.h
Comment (garbear) [inline xbmc/windowing/gbm/WinSystemGbm.h:89]: Can you document that `nullptr` means to unset HDR?

File: xbmc/windowing/gbm/WinSystemGbm.cpp
Comment (Kwiboo) [inline xbmc/windowing/gbm/WinSystemGbm.cpp:328]: `COLOR_ENCODING` and `COLOR_RANGE` should be set regardless if `HDR_OUTPUT_METADATA` is supported or not, and the color encoding/range from video picture may also not be applicable for the gui plane.

**Review by Kwiboo — DISMISSED** [review]

File: xbmc/windowing/gbm/WinSystemGbm.cpp
Comment (lrusak) [reply xbmc/windowing/gbm/WinSystemGbm.cpp:328]: They are still set independently from the HDR metadata. I'm also not sure how it would be different for video/gui plane?

Comment (Kwiboo) [reply xbmc/windowing/gbm/WinSystemGbm.cpp:328]: This code will not always run because of `if (!settings->GetBool(SETTING_WINSYSTEM_IS_HDR_DISPLAY))` few lines up.

The color plane properties is used to inform the drm driver what color properties a non RGB plane has.
The gui plane color properties should match the EGL context, I think kodi always use an RGB plane as gui plane if I am not mistaken (I am unsure how HDR is handled in EGL).

**Comment by lrusak** [comment]
jenkins build this please

File: xbmc/windowing/gbm/WinSystemGbm.h
Comment (garbear) [reply xbmc/windowing/gbm/WinSystemGbm.h:60]: I'd have squashed. You live life on the edge. All should look to you.

**Review by garbear — APPROVED** [review]

**Comment by lrusak** [comment]
jenkins build this please

**Comment by bertin0** [comment]
> Tested on Pi4. HDR colours correct when "Use display HDR" is enabled and incorrect when disabled.

So is it possible to playback HDR content on linux (eg raspberry pi 4) by running in drm mode?

**Comment by popcornmix** [comment]
> So is it possible to playback HDR content on linux (eg raspberry pi 4) by running in drm mode?

Yes, this has been supported in LibreELEC (as one example) for a long long time.

**Comment by lrusak** [comment]
Anyone have an opinion on dismissing a stale review? I can't merge this while there are requested changes (that have been addressed) by @Kwiboo. AFAIK he is busy and inactive (but may see this ping).

**Comment by lrusak** [comment]
Tested and verified working. Some log snippets:
```
2022-08-03 20:47:20.980 T:976      INFO <general>: ffmpeg[0x5742d30]:   Stream #0:0(und): Video: hevc (Main 10) (hvc1 / 0x31637668), yuv420p10le(tv, bt2020nc/bt2020/smpte2084), 3840x2160 [SAR 1:1 DAR 16:9], 74462 kb/s, 59.94 fps
...
2022-08-03 20:47:26.301 T:980     DEBUG <general>: VAAPI - open decoder
2022-08-03 20:47:26.301 T:980     DEBUG <general>: libva info: VA-API version 1.15.0
2022-08-03 20:47:26.302 T:980     DEBUG <general>: libva info: Trying to open /usr/lib/dri/iHD_drv_video.so
2022-08-03 20:47:26.302 T:980     DEBUG <general>: libva info: Found init function __vaDriverInit_1_15
2022-08-03 20:47:26.304 T:980     DEBUG <general>: libva info: va_openDriver() returns 0
2022-08-03 20:47:26.304 T:980     DEBUG <general>: VAAPI - initialize version 1.15
2022-08-03 20:47:26.304 T:980     DEBUG <general>: VAAPI - driver in use: Intel iHD driver for Intel(R) Gen Graphics - 22.5.1 (56e30c407c)
...
2022-08-03 20:47:26.321 T:980     DEBUG <general>: CDVDVideoCodecFFmpeg - Updated codec: ff-hevc-vaapi
...
2022-08-03 20:47:26.766 T:980     DEBUG <general>: CRenderManager::Configure - change configuration. 3840x2160. display: 3840x2160. framerate: 59.94.
2022-08-03 20:47:26.797 T:734     DEBUG <general>: LinuxRendererGLES::Configure: fps: 59.940
2022-08-03 20:47:26.797 T:734     DEBUG <general>: CWinSystemGbm::SetHDR - display_primaries[0].x: 50000
2022-08-03 20:47:26.797 T:734     DEBUG <general>: CWinSystemGbm::SetHDR - display_primaries[0].y: 50000
2022-08-03 20:47:26.797 T:734     DEBUG <general>: CWinSystemGbm::SetHDR - display_primaries[1].x: 50000
2022-08-03 20:47:26.797 T:734     DEBUG <general>: CWinSystemGbm::SetHDR - display_primaries[1].y: 50000
2022-08-03 20:47:26.797 T:734     DEBUG <general>: CWinSystemGbm::SetHDR - display_primaries[2].x: 50000
2022-08-03 20:47:26.797 T:734     DEBUG <general>: CWinSystemGbm::SetHDR - display_primaries[2].y: 50000
2022-08-03 20:47:26.797 T:734     DEBUG <general>: CWinSystemGbm::SetHDR - white_point.x: 50000
2022-08-03 20:47:26.797 T:734     DEBUG <general>: CWinSystemGbm::SetHDR - white_point.y: 50000
2022-08-03 20:47:26.797 T:734     DEBUG <general>: CWinSystemGbm::SetHDR - max_display_mastering_luminance: 1
2022-08-03 20:47:26.797 T:734     DEBUG <general>: CWinSystemGbm::SetHDR - min_display_mastering_luminance: 1000
2022-08-03 20:47:26.797 T:734     DEBUG <general>: CWinSystemGbm::SetHDR - max_cll: 0
2022-08-03 20:47:26.797 T:734     DEBUG <general>: CWinSystemGbm::SetHDR - max_fall: 0
2022-08-03 20:47:26.797 T:734     DEBUG <general>: LinuxRendererGLES::Configure: HDR passthrough: on
```

**Pull request merged**
